pulsar-testclient/src/main/java/org/apache/pulsar/testclient/PerformanceTransaction.java

```diff
@@ -0,0 +1,637 @@
+        if (arguments.partitions != null) {
+            PulsarAdminBuilder clientBuilder = PulsarAdmin.builder()
+                    .serviceHttpUrl(arguments.adminURL);
+            try (PulsarAdmin client = clientBuilder.build()) {
+                for (String topic : arguments.producerTopic) {
+                    log.info("Creating  produce partitioned topic {} with {} partitions", topic, arguments.partitions);
+                    try {
+                        client.topics().createPartitionedTopic(topic, arguments.partitions);
+                    } catch (PulsarAdminException.ConflictException alreadyExists) {
+                        if (log.isDebugEnabled()) {
+                            log.debug("Topic {} already exists: {}", topic, alreadyExists);
+                        }
+                        PartitionedTopicMetadata partitionedTopicMetadata =
+                                client.topics().getPartitionedTopicMetadata(topic);
+                        if (partitionedTopicMetadata.partitions != arguments.partitions) {
+                            log.error(
+                                    "Topic {} already exists but it has a wrong number of partitions: {}, expecting {}",
+                                    topic, partitionedTopicMetadata.partitions, arguments.partitions);
+                            PerfClientUtils.exit(-1);
+                        }
+                    }
+                }
+            }
+        }
+
+        PulsarClient client =
+                PulsarClient.builder().enableTransaction(arguments.isEnableTransaction)
+                        .serviceUrl(arguments.serviceURL)
+                        .connectionsPerBroker(arguments.maxConnections)
+                        .statsInterval(0, TimeUnit.SECONDS)
+                        .ioThreads(arguments.ioThreads)
+                        .build();
+
+        ExecutorService executorService = Executors.newFixedThreadPool(arguments.numTestThreads);
+        Semaphore testThreadLimit = new Semaphore(arguments.numTestThreads);
+
+        long startTime = System.nanoTime();
+        long testEndTime = startTime + (long) (arguments.testTime * 1e9);
+        Runtime.getRuntime().addShutdownHook(new Thread(() -> {
+            if (arguments.isEnableTransaction) {
+                printTxnAggregatedThroughput(startTime);
+            } else {
+                printAggregatedThroughput(startTime);
+            }
+            printAggregatedStats();
+        }));
+
+        // start perf test
+        AtomicBoolean executing = new AtomicBoolean(true);
+        new Thread(() -> {
+            RateLimiter rateLimiter = null;
+            if(arguments.openTxnRate != 0){
+                rateLimiter = RateLimiter.create( (double)arguments.openTxnRate / arguments.numTestThreads);
+            }
+            for(int i = 0; i < arguments.numTestThreads; i++) {
+                if (testThreadLimit.tryAcquire() && (rateLimiter == null || rateLimiter.tryAcquire())) {
```

Comment:
does this rate should use in openTxn, not here.

Reply:
Here has a conflict : a test thread will only open a transaction at one time. Maybe this should be delete?